src/hierarchies/base_hierarchy.h

```diff
@@ -29,101 +29,312 @@
-                                         const Eigen::VectorXd &covariate,
-                                         bool add) = 0;
-  //!
-  virtual void save_posterior_hypers() = 0;
-  //!
-  virtual void initialize_hypers() = 0;
-  //!
-  virtual void create_empty_prior() = 0;
 
  public:
+  virtual ~AbstractHierarchy() = default;
+  virtual std::shared_ptr<AbstractHierarchy> clone() const = 0;
+  virtual bayesmix::HierarchyId get_id() const = 0;
+
   //! Adds a datum and its index to the hierarchy
-  void add_datum(const int id, const Eigen::VectorXd &datum,
-                 const bool update_params = false,
-                 const Eigen::VectorXd &covariate = Eigen::VectorXd(0));
+  virtual void add_datum(
+      const int id, const Eigen::VectorXd &datum,
+      const bool update_params = false,
+      const Eigen::VectorXd &covariate = Eigen::VectorXd(0)) = 0;
   //! Removes a datum and its index from the hierarchy
-  void remove_datum(const int id, const Eigen::VectorXd &datum,
-                    const bool update_params = false,
-                    const Eigen::VectorXd &covariate = Eigen::VectorXd(0));
+  virtual void remove_datum(
+      const int id, const Eigen::VectorXd &datum,
+      const bool update_params = false,
+      const Eigen::VectorXd &covariate = Eigen::VectorXd(0)) = 0;
   //! Deletes all data in the hierarchy
-  virtual void clear_data() = 0;
-  //!
   virtual void initialize() = 0;
-  void check_prior_is_set();
-  //! Returns true if the hierarchy models multivariate data
+
   virtual bool is_multivariate() const = 0;
-  //! Returns true if the hierarchy has covariates i.e. is a dependent model
   virtual bool is_dependent() const { return false; }
-  //! Returns true if the hierarchy is conjugate i.e. has a marginal lpdf
   virtual bool is_conjugate() const { return true; }
   //!
   virtual void update_hypers(
       const std::vector<bayesmix::MarginalState::ClusterState> &states) = 0;
 
-  // DESTRUCTOR AND CONSTRUCTORS
-  virtual ~BaseHierarchy() = default;
-  BaseHierarchy() = default;
-  virtual std::shared_ptr<BaseHierarchy> clone() const = 0;
-
   // EVALUATION FUNCTIONS FOR SINGLE POINTS
   //! Evaluates the log-likelihood of data in a single point
   virtual double like_lpdf(
       const Eigen::RowVectorXd &datum,
-      const Eigen::RowVectorXd &covariate = Eigen::VectorXd(0)) const = 0;
+      const Eigen::RowVectorXd &covariate = Eigen::VectorXd(0)) = 0;
+
   //! Evaluates the log-marginal distribution of data in a single point
-  virtual double marg_lpdf(
-      const bool posterior, const Eigen::RowVectorXd &datum,
-      const Eigen::RowVectorXd &covariate = Eigen::VectorXd(0)) const = 0;
+  virtual double prior_pred_lpdf(
+      const Eigen::RowVectorXd &datum,
+      const Eigen::RowVectorXd &covariate = Eigen::VectorXd(0)) = 0;
+
+  virtual double conditional_pred_lpdf(
+      const Eigen::RowVectorXd &datum,
+      const Eigen::RowVectorXd &covariate = Eigen::VectorXd(0)) = 0;
```

Comment:
Why are these methods no longer `const`?

Reply:
I will revert to it, it just needs a different kind of casting in `BaseHierarchy`